contrib/run_task.sh

```diff
@@ -0,0 +1,230 @@
+#!/bin/env bash
+
+set -ex
+
+# Make all cargo invocations verbose.
+export CARGO_TERM_VERBOSE=true
+
+cargo --version
+rustc --version
+
+main() {
+    crate="$1"
+    task="$2"
+
+    cd "$crate"
+
+    # Building the fuzz crate is more-or-less just a sanity check.
+    if [ "$crate" == "fuzz" ]
+    then
+        cargo --locked build
+        exit 0
+    fi
+
+    # Every crate must define EXAMPLES.
+    . contrib/test_vars.sh || exit 1
+
+    case $task in
+	test)
+	    do_test
+	    ;;
+
+	feature_matrix)
+	    do_feature_matrix
+	    ;;
+
+	lint)
+	    do_lint
+	    ;;
+
+	dup_deps)
+	    do_dup_deps
+	    ;;
+
+	docs)
+            build_docs_with_stable_toolchain
+	    ;;
+
+	docsrs)
+            build_docs_with_nightly_toolchain
+	    ;;
+
+	wasm)
+	    do_wasm
+	    ;;
+
+	asan)
+	    do_asan
+	    ;;
+
+	bench)
+	    do_bench
+	    ;;
+
+	schemars)
+	    do_schemars
+	    ;;
+	*)
+	    echo "Error: unknown task $task" >&2
+	    exit 1
+	    ;;
+    esac
+}
+
+do_test() {
+    # Use the current (recent/minimal) lock file.
+    local cargo="cargo --locked"
+
+    # Defaults / sanity checks
+    $cargo build
+    $cargo test
+
+    for example in $EXAMPLES; do
+	name="$(echo "$example" | cut -d ':' -f 1)"
+	features="$(echo "$example" | cut -d ':' -f 2)"
+	$cargo run --example "$name" --features="$features"
+    done
+
+    if [ -e ./contrib/extra_tests.sh ];
+    then
+	./contrib/extra_tests.sh
+    fi
+}
+
+# Each crate defines its own feature matrix test so feature combinations
+# can be better controlled.
+do_feature_matrix() {
+    local cargo="cargo --locked"
+
+    $cargo build --no-default-features
+    $cargo test --no-default-features
+
+    # All crates have a "std" feature.
+    loop_features "std" "$FEATURES_WITH_STD"
+
+    # All but `bitcoin` crate have an "alloc" feature, this tests it
+    # along with any other features that should work with "std".
+    if [ -n "$FEATURES_WITHOUT_STD" ]
+       then
+           loop_features "" "$FEATURES_WITHOUT_STD"
+    fi
+}
+
+# Build with each feature as well as all combinations of two features.
+#
+# Usage: loop_features "std" "this-feature that-feature other"
+loop_features() {
+    local use="$1"
+    local features="$2"
+    local cargo="cargo --locked"
+
+    # All the provided features including $use
+    $cargo build --no-default-features --features="$use $features"
+    $cargo test --no-default-features --features="$use $features"
+
+    read -r -a outer <<< "$features"
+    local len="${#outer[@]}"
+
+    if (( len > 1 )); then
+        for ((i = 0 ; i < len ; i++ ));
+        do
+            $cargo build --features="$use ${outer[i]}"
+            $cargo test --features="$use ${outer[i]}"
+
+            if (( i < len - 1 )); then
+               inner=("${outer[@]:i+1}")
+               for ((j = 0 ; j < ${#inner[@]} ; j++ ));
```

Comment:
Why not simply `j = i + 1; j < len; j++` without the if and having to make another array?

Reply:
Laughing at myself, looks like I wrote first year code - used as suggested.